node/src/components/small_network.rs

```diff
@@ -953,6 +884,67 @@ where
                     Effects::new()
                 }
             }
+            Event::ContractRuntimeAnnouncement(
+                ContractRuntimeAnnouncement::LinearChainBlock(_)
+                | ContractRuntimeAnnouncement::StepSuccess { .. },
+            ) => Effects::new(),
+            Event::ContractRuntimeAnnouncement(
+                ContractRuntimeAnnouncement::UpcomingEraValidators {
+                    era_that_is_ending,
+                    mut upcoming_era_validators,
+                },
+            ) => {
+                if era_that_is_ending < self.active_era {
+                    debug!("ignoring past era end announcement");
+                } else if era_that_is_ending != self.active_era {
+                    debug!("received future era end announcement, clearing limiters");
+                    self.active_era = era_that_is_ending + 1;
+                    self.active_era_validators = Default::default();
+
+                    self.incoming_limiter
+                        .update_validators(Default::default(), Default::default());
+                    self.outgoing_limiter
+                        .update_validators(Default::default(), Default::default());
+                } else {
+                    // At this point we can rely on the `active_era` being the one that is ending.
+                    self.active_era = era_that_is_ending + 1;
+
+                    // Initialize the next set of active validators with the currently active one.
+                    let mut active_validators: HashSet<PublicKey> =
+                        mem::take(&mut self.active_era_validators);
+
+                    self.active_era_validators = upcoming_era_validators
+                        .remove(&self.active_era)
+                        .unwrap_or_default()
+                        .into_keys()
+                        .collect();
+
+                    if self.active_era_validators.is_empty() {
+                        warn!("received an empty set of active era validators");
+                    }
+
+                    active_validators.extend(self.active_era_validators.iter().cloned());
+
+                    let upcoming_validators: HashSet<PublicKey> = upcoming_era_validators
+                        .remove(&(self.active_era + 1))
+                        .unwrap_or_default()
+                        .into_keys()
+                        .collect();
+
```

Comment:
```suggestion
                    // Initialize the next set of active validators with the currently active one.
                    let mut active_validators: HashSet<PublicKey> =
                        mem::take(&mut self.active_era_validators);

                    self.active_era_validators = upcoming_era_validators
                        .remove(&self.active_era)
                        .unwrap_or_default()
                        .into_keys()
                        .collect();

                    if self.active_era_validators.is_empty() {
                        warn!("received an empty set of active era validators");
                    }

                    let active_validators = upcoming_era_validators
                        .remove(&(era_that_is_ending + 1))
                        .unwrap_or_default()
                        .into_keys()
                        .collect();;

                    let upcoming_validators: HashSet<PublicKey> = upcoming_era_validators
                        .remove(&(era_that_is_ending + 2))
                        .unwrap_or_default()
                        .into_keys()
                        .collect();
```

Looks like we don't need `self.active_validators`...

Reply:
There is no `self.active_validators`, only `self.active_era_validators` and `active_validators`.

`self.active_era_validators` keeps the validators of the currently active era, because we want to include them in the preferred peers set for one era after their active era ends. This, of course, assumes that they are not included in the announcement.

`active_validators` is the union of this previous set of validators and the very-soon-to-be-actually-active validators from the announcement and is passed to the limiters.

I believe we need both.

Reply:
Obsolete with c5720c575